setup.py

```diff
@@ -15,6 +15,7 @@
     "dill",
     "pandas",
     "pyyaml",
+    "click",
 ]
```

Comment:
General remark about the requirements file - shouldn't be there a specific version of the package? Or should always the latest version be installed and in case something doesn't work with the latest version then we would use the specific version?

Reply:
Great question!

TL;DR: You want to play it loose with requirements for libraries (which `mllaunchpad` is), and strict for applications (which `mllaunchpad` is not)

In this case, for a *library* (which `mllaunchpad` is), not specifying specific versions is okay. Because our users will have to manage a bunch of their own dependencies (=requirements), too, we want to give them as much flexibility as possible. Otherwise, if we specified for example `pandas==1.0.1` and the library's users had to use a pre-1.0 version of `pandas` to be compatible with some other library they want to use, this would cause them problems.

Not specifying a version is not the same as saying "use the newest version available". While in practice, of course, `pip` tries to get the newest versions that are available, we leave this open, so that there's wiggle room to find compatible versions. Some tools (not `pip` ([at least](https://github.com/pypa/pip/issues/988) [not yet](https://twitter.com/di_codes/status/1193980331004743680)), but e.g. `pip-tools`, `pipenv`, `poetry` and `dephell`) do proper dependency resolution, searching for versions that satisfy all dependencies (also inter-dependency).

That said, as soon as we learn from actual users that `mllaunchpad` does not work with specific versions of its dependencies, such as very old versions of pandas, etc., there's no harm in adding these `>=[version]` contstraints to `setup.py` then. I'd rather react if something doesn't work, than make a choice now that is maybe ill-informed and excludes some users. It would be a lot of work to make a well-informed choice on this (to try out old dependency versions and see if everything still works). Maybe when we have good unit test coverage, we might want to run a matrix check on the dependency versions to fill in this information.

As a side note, one practice I see a lot in libraries recently and explicitly *don't* want to adopt is that some libraries just generally exclude the next major version of all their dependencies. I get that they want to make sure that nothing breaks when some dependency has a breaking change, because this would be out of their control. But this is only acceptable if they have some daily check running that automatically tests new versions of all dependencies for compatibility and then auto-deploys a new version of their package to pypi with the updated dependency specifications. Otherwise, they just play it safe on the backs of their users, forbidding installs of their library in settings where they would perfectly, creating even more of a dependency hell for their users.

For *applications* on the other hand (which `mllaunchpad` is not), you definitely want to freeze/pin your dependencies to specific versions, for reproducibility of your deployments. For example, data scientists who use `mllaunchpad` to deploy an API should absolutely pin *their* project's dependencies.